### PR TITLE
Adding option to force all traffic to HTTPS

### LINKED ISSUE #4871: Add option to automatically redirect all requests to HTTPS
On an ssl-enabled Heroku instance, both http and https are accessible, which means that users may inadvertently login over http. A redirect from http to https would be appropriate here.

This is similar to #1356, but then on Heroku instead of AWS. On Heroku one cannot configure a webserver to perform the redirect. Basically, if `X-Forwarded-Proto` is present and not `https` and the site url uses https, a redirect is needed.

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by robdaemon**
Adds a boolean setting to force redirects to HTTPS (defaults to false)
and a integer setting for the SSL port number.

Resolves #4871

### CODE REVIEW

**Comment by robdaemon** [comment]
Do not merge yet, still need to write tests, but wanted to push up current state EOD

**Comment by camsaul** [comment]
I tuned the CodeCov stuff a bit so try merging master and it should fix the CI check failures.

For this PR we need to make sure it works with reverse proxies like nginx or ELBs that handle SSL termination. Take a look at `metabase.middleware.session/https-request?` -- we have a function that takes in to account `X-Fowarded-Proto`, `X-Forwarded-Ssl`, `Front-End-Https`, and a variety of other headers to try to determine if the original request was HTTPS (not just the request ultimately forwarded to Ring)

**Comment by robdaemon** [comment]
Sounds good, will make those changes now.

Also, do we have any URLs we want to exclude from this redirect, like a healthcheck or status?

**Comment by camsaul** [comment]
@robdaemon ooo, yeah. We definitely don't want to force the health endpoint to be HTTPS, since that could break ELB or other load balancers . The health check endpoint is called `GET /api/health` so you can carve out an exception for that. If you look at https://github.com/metabase/metabase/blob/master/src/metabase/middleware/log.clj#L177-L181 we're doing something similar for logging -- we don't log calls to the health endpoint or the endpoint that fetches logs

File: project.clj
Comment (camsaul) [inline project.clj:136]: I didn't know this existed. Does it do what we want?

Comment (robdaemon) [reply project.clj:136]: not everything - it can't do the conditions we want, but it's still useful

**Comment by camsaul** [comment]
@robdaemon if you want to try adding the UI for the setting yourself it's pretty straightforward. Settings defined with `defsetting` are already viewable/setable by admins with the relevant REST API endpoints. So all you have to do now is put an entry in `frontend/src/metabase/admin/settings/selectors.js` to tell the FE client where to render the widget for changing this setting. @mazameli suggested putting it after `site-locale` in the `General` tab. So you can add an entry [here](https://github.com/metabase/metabase/blob/09a868be79306b0b3e9bd4a5f9cb19bf94c68e90/frontend/src/metabase/admin/settings/selectors.js#L69-L77). 

If you run `yarn run build-hot` the FE will automatically re-build and update when you make changes so it should show up after you add it. You can add an entry for the SSL port and then add a `getHidden` function so the port setting only shows if redirect to SSL is enabled.

I think @tlrobinson's suggestion in #4871 to check that HTTPS actually works before letting someone enable it is a good idea. I'm not an expert on that code but I think you can just add that function to `validations` for the entry. @tlrobinson what's the best place to add that code?

**Comment by robdaemon** [comment]
I'll give that a shot in the morning!

**Comment by camsaul** [comment]
I was just thinking about this...

We need to make sure we forward to [`X-Forwarded-Host`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Host) instead of `Host` if it's present so this works properly with reverse proxies. It doesn't look like the Ring SSL lib does that

https://github.com/ring-clojure/ring-ssl/blob/master/src/ring/middleware/ssl.clj#L51
https://github.com/ring-clojure/ring/blob/master/ring-core/src/ring/util/request.clj#L5

**Comment by camsaul** [comment]
`metabase.public-settings/site-url` might be better than looking at `X-Forwared-Host`/`Host` because that's something the user can actually control, so they can make sure they set it to the correct value

**Review by camsaul — APPROVED** [review]
Cool, this works for me. Now that we're using `site-url` I think the JS-side validation is less important.

**Comment by mazameli** [comment]
I'd like to make sure we do a UX review for anything user-facing like this before merging, generally speaking. In testing it just now, I did in fact manage to accidentally lock myself out of my local instance.

There are are couple things about this that feel off:

- Were we not able to implement Tom's suggestion about verifying that HTTPS is available? I would strongly like us to do whatever we can to prevent users from accidentally locking themselves out of the instance.
- It appears that the implemented solution requires the user to type HTTPS for the site URL's protocol in order for this setting to take effect. I'm unclear if this would be necessary if the above were implemented — but this current state of affairs isn't great since this lets me potentially turn on a setting that doesn't do anything. Could we look at the value of the Site URL input, and if it doesn't begin with `https:`, prevent the toggle from being clicked on? (Alternatively, we could _change_ the input from `http` to `https` if the setting is toggled on, but I'm guessing there are good reasons we wouldn't want to do that if HTTPS isn't actually available.)

**Comment by robdaemon** [comment]
@mazameli we can revert the merge if you feel like there's more work to be done here?

**Comment by mazameli** [comment]
I'm not sure what the preferred course of action is on this re: reverting vs. doing a follow-on pr, but yeah, there's some cleanup work to be done. We can sync up in the AM if that sounds good.

**Pull request merged**
